_blog/2022-12-23-procedural-macros.mdx

```diff
@@ -0,0 +1,417 @@
+```javascript
+const stringToFloat32 = (string, offset) => {
+	const u16 = new Uint16Array([
+		string.charCodeAt(offset),
+		string.charCodeAt(offset + 1),
+	]).buffer;
+	return new Float32Array(u16)[0];
+};
+
+function arrayFromString(string, fields) {
+	let i = 0;
+	const entries = stringToFloat32(string, i);
+    i += 2;
+	const array = [];
+	for (let arrayIndex = 0; arrayIndex < entries; arrayIndex++) {
+        const object = {};
+		for (const { name, ty } of fields) {
+			let value;
+			switch (ty) {
+				case "string":
+					const length = stringToFloat32(string, i);
+					i += 2;
+					value = String.fromCharCode(
+						...Array.from({ length }, (_, j) => string.charCodeAt(j + i)),
+					);
+					i += length;
+					break;
+				case "number":
+					value = stringToFloat32(string, i);
+					i += 2;
+					break;
+				case "boolean":
+					value = string.charCodeAt(i) === 0;
+					i++;
+					break;
+			}
+			Reflect.set(object, name, value);
+		}
+        array.push(object);
+	}
+	return array;
+}
+```
+
+Here we use another part of reflection `Reflect.set` which allows us to set a property of an existing object based on a string key (`name`).
+### Problems with reflection in JavaScript
+The `arrayFromString` function assumes that the caller has passed a standard array where every object has the same type. `Reflect.get` will fail at runtime as if the property doesn't not exist `null` will be returned and there is a catch to throw an `Error`, so nothing unsafe. The downside is the both the dynamic property lookup and null check is expensive at runtime. 
+
+One way to catch property errors ahead of time is if we had used an extension to JS like TypeScript, which can type the parameters of the function and check the arguments of callers.
+## Writing a Rust procedural macro
+If you are not in a Rust project/workspace already you can create one with the `cargo new *name*` command. Before we start generating code we should declare a trait as a target for our macros output.
+### The BinarySerializable trait
+In our current create, we need to create a `trait` that will be describe the requirements for serializing and deserializing objects. It will be called `BinarySerializable`. Serialization will require adding data to a buffer. Deserialization will require pulling from an iterator (which iterates over bytes of a serialized buffer) and producing a term of `Self`. *We will assume the format is well formed and panic at runtime rather than proper handling with `Result` when deserializing.*
+```rust
+pub trait BinarySerializable {
+    fn serialize(self, buf: &mut Vec<u8>);
+
+    fn deserialize<I: Iterator<Item = u8>>(iter: &mut I) -> Self;

```

Comment:
I won't expect to find a `deserialize` method in a `Serializable` trait. Maybe two traits are needed? Or this trait name has to be something else?

Reply:
Hmm yes. Serde does this using two traits. It would be a lot tidier to have one trait and thus one proc-macro. So maybe the name should be `TwoWaySerializable` or `BinarySerializableAndDeserializable`?

Reply:
Or just `Binary`